packages/phone-number-privacy/common/src/interfaces/requests.ts

```diff
@@ -150,35 +146,156 @@ export type DomainQuotaStatusRequest<
-  D extends Domain = Domain,
-  O extends DomainOptions = D extends KnownDomain ? KnownDomainOptions<D> : never
-> = OmitIfNever<
-  {
-    /** Domain specification. Selects the PRF domain and rate limiting rules. */
-    domain: D
-    /** Domain-specific options. */
-    options: O
-    /** Client-specified session ID. */
-    sessionID: EIP712Optional<string>
-  },
-  'options'
->
+export type DisableDomainRequest<D extends Domain = Domain> = {
+  /** Domain specification. Selects the PRF domain and rate limiting rules. */
+  domain: D
+  /** Domain-specific options. */
+  options: DomainOptions<D>
+  /** Client-specified session ID. */
+  sessionID: EIP712Optional<string>
+}
 
 /** Union type of Domain API requests */
-export type DomainRequest<
-  D extends Domain = Domain,
-  O extends DomainOptions = D extends KnownDomain ? KnownDomainOptions<D> : never
-> =
-  | DomainRestrictedSignatureRequest<D, O>
-  | DomainQuotaStatusRequest<D, O>
-  | DisableDomainRequest<D, O>
+export type DomainRequest<D extends Domain = Domain> =
+  | DomainRestrictedSignatureRequest<D>
+  | DomainQuotaStatusRequest<D>
+  | DisableDomainRequest<D>
+
+// NOTE: Next three functions are a bit repetitive. An attempt was made to combine them, but the
+// type signature got quite complicated. Feel free to attempt it if you are motivated.
+
+/** Parameterized schema for checking unknown input against DomainRestrictedSignatureRequest */
+export function domainRestrictedSignatureRequestSchema<D extends Domain = Domain>(
+  domain: t.Type<D>
+): t.Type<DomainRestrictedSignatureRequest<D>> {
+  // The schema defined here does most of the work, but does not guarantee consistency between the
+  // domain and options fields. We wrap the schema below to add a consistency check.
+  const schema = t.strict({
+    domain,
+    options: t.unknown,
+    blindedMessage: t.string,
+    sessionID: eip712OptionalSchema(t.string),
+  })
+
+  const validation = (
+    unk: unknown,
+    ctx: t.Context
+  ): t.Validation<DomainRestrictedSignatureRequest<D>> =>
+    pipe(
+      schema.validate(unk, ctx),
+      chain((value: t.TypeOf<typeof schema>) => {
+        if (isSequentialDelayDomain(value.domain)) {
+          const either = SequentialDelayDomainOptionsSchema.validate(value.options, ctx)
+          if (isRight(either)) {
+            return t.success(value as DomainRestrictedSignatureRequest<D>)
+          }
+
+          return t.failure(unk, ctx, 'options type does not match domain type')
+        }
+
+        // canary provides a compile-time check that all subtypes of Domain have branches. If a case
+        // was missed, then an error will report that domain cannot be assigned to type `never`.
+        const canary = (x: never) => x
+        canary(value.domain)
+        throw new Error('Implementation error: validated domain is not of any known type')
+      })
+    )
+
+  return new t.Type<DomainRestrictedSignatureRequest<D>, DomainRestrictedSignatureRequest<D>>(
```

Comment:
Is the DomainRestrictedSignatureRequest<D> type parameter supposed to be passed in twice here?

Reply:
It is, ya. It's to specify that both the "main" type and encoder output type of the `io-ts` codec are of type `DomainRestrictedSignatureRequest`. It is usually sufficient to specify only the first arg, which is the main type for the codec used as the output of `decode` and the input of encoding, but in this case the compiler needed some clarification, so encoder output arg is also provided.

https://github.com/gcanti/io-ts/blob/master/src/index.ts#L146

This hits on one of the things that mildly confusing things about using `io-ts` in my experience so far. Namely that it uses the language of "encoding" and "decoding" although for out purposes it's usually the case that we have an object already (often decoded from JSON by another framework) so what we are actually doing is "decoding" from `unknown` to the type we want. In our use case, this would be more naturally described as a type check, because no data manipulation is happening, but it makes more sense in the case that you are additionally using `io-ts` to take a serialized object (e.g. as a `string`) and decoding all the way to the type you want.